experimental/transliterate/src/compile/mod.rs

```diff
@@ -39,43 +38,122 @@ impl Direction {
-/// that BCP47 ID. For unknown legacy IDs, the output is given by this function.
-pub fn legacy_id_to_internal_id(source: &str, target: &str, variant: Option<&str>) -> String {
-    // TODO(#3891): Decide representation for unknown BCP47 IDs
-    let mut id = format!("x-{source}-{target}");
-    if let Some(variant) = variant {
-        id.push('-');
-        id.push_str(variant);
-    }
-    // normalizing to ASCII lowercase to avoid duplicate dependencies like Any-null and Any-Null
-    id.make_ascii_lowercase();
-    id
+/// # Example
+/// ```
+/// use icu_transliterate::{RuleCollection, Transliterator};
+///
+/// let mut collection = RuleCollection::default();
+/// collection.register_source(
+///     &"de-t-de-d0-ascii".parse().unwrap(),
+///     r#"
+///       $AE = [Ä {A \u0308}];
+///       $OE = [Ö {O \u0308}];
+///       $UE = [Ü {U \u0308}];
+///  
+///       [ä {a \u0308}] → ae;
+///       [ö {o \u0308}] → oe;
+///       [ü {u \u0308}] → ue;
+///  
+///       {$AE} [:Lowercase:] → Ae;
+///       {$OE} [:Lowercase:] → Oe;
+///       {$UE} [:Lowercase:] → Ue;
+///  
+///       $AE → AE;
+///       $OE → OE;
+///       $UE → UE;
+///  
+///       ::Any-ASCII;
+///     "#.to_string(),
+///     // legacy alias, doesn't actually do anything in this case
+///     ["de-ASCII"],
+///     false,
+///     true,
+/// );
+/// collection.register_source(
+///     &"und-t-und-latn-d0-ascii".parse().unwrap(),
+///     "# ... ".to_string(),
+///     // the previous transliterator refers to this one by the `Any-ASCII` ID,
+///     // so it's important that the alias is listed here.
+///     ["Latin-Ascii", "Any-ASCII"],
+///     false,
+///     true,
+/// );
+///
+/// Transliterator::try_new_unstable("de-t-de-d0-ascii".parse().unwrap(), &collection.as_provider()).unwrap();
+///
+pub struct RuleCollection {
+    id_mapping: HashMap<String, Locale>, // alias -> bcp id
+    data: RwLock<HashMap<DataLocale, (String, bool, bool)>>, // locale -> source/reverse/visible
+    cache: RwLock<HashMap<DataLocale, DataResponse<TransliteratorRulesV1Marker>>>,
 }
 
-impl RuleBasedTransliterator<'static> {
-    /// Parse a rule based transliterator definition into a `TransliteratorDataStruct`.
-    ///
-    /// See [UTS #35 - Transliterators](https://unicode.org/reports/tr35/tr35-general.html#Transforms) for more information.
-    ///
-    /// The `direction` [`Direction`] argument
-    /// determines which of the returned options is populated. The first option will be populated
-    /// if the direction is [`Forward`](Direction::Forward), the second if the direction is
-    /// [`Reverse`](Direction::Reverse), and both if the direction is [`Both`](Direction::Both).
+impl RuleCollection {
+    /// Add a new transliteration source to the collection.
+    pub fn register_source<'a>(
+        &mut self,
+        id: &icu_locid::Locale,
+        source: String,
+        aliases: impl IntoIterator<Item = &'a str>,
+        reverse: bool,
+        visible: bool,
```

Comment:
Nit: Avoid boolean positional arguments because they are not readable at the call site; prefer enums